### DISCUSSION TITLE
Help with First Person Shooter

### DISCUSSION THREAD

**Discussion by N8python** [enable3d/enable3d#101] [Help / Questions] [answered]
I was experimenting with FPS example and I noticed that the rifle could be offset on all axes position-wise, in the following code:
```js
const raycaster = new THREE.Raycaster()
            // x and y are normalized device coordinates from -1 to +1
            raycaster.setFromCamera({ x: 0.6 - this.move.x, y: -0.8 - this.move.y }, this.third.camera)
            const pos = new THREE.Vector3()
            pos.copy(raycaster.ray.direction)
            pos.multiplyScalar(0.8 + this.move.z)
            pos.add(raycaster.ray.origin)

            this.rifle.position.copy(pos)
            this.rifle.rotation.copy(this.third.camera.rotation)
```
What I was wondering is if there was a way to do the same for rotation - is there any math/function I could use to rotate the rifle with respect to the camera? I tried just modifying the angle, but that causes the sword to start spinning in weird ways - is there something I'm missing?

**Comment by vitalspace** [comment]
> I was experimenting with FPS example and I noticed that the rifle could be offset on all axes position-wise, in the following code:
> 
> ```js
> const raycaster = new THREE.Raycaster()
>             // x and y are normalized device coordinates from -1 to +1
>             raycaster.setFromCamera({ x: 0.6 - this.move.x, y: -0.8 - this.move.y }, this.third.camera)
>             const pos = new THREE.Vector3()
>             pos.copy(raycaster.ray.direction)
>             pos.multiplyScalar(0.8 + this.move.z)
>             pos.add(raycaster.ray.origin)
> 
>             this.rifle.position.copy(pos)
>             this.rifle.rotation.copy(this.third.camera.rotation)
> ```
> 
> What I was wondering is if there was a way to do the same for rotation - is there any math/function I could use to rotate the rifle with respect to the camera? I tried just modifying the angle, but that causes the sword to start spinning in weird ways - is there something I'm missing?

check https://github.com/enable3d/enable3d-website/blob/master/src/examples/first-person-shooter.html

**Reply by N8python** [reply]
Yes I looked at that. The problem is is that that file only shows how to offset the weapon, not how to rotate it.

**Comment by yandeu** [comment, marked as answer]
Try the add the following just after the code snippet you posted:

```js
this.rifle.rotateX(Math.PI / 8)
this.rifle.rotateY(Math.PI / 8)
this.rifle.rotateZ(Math.PI / 8)
```

**Reply by N8python** [reply]
Thank you so much. This worked great. It's awesome that you are so available.
